### PR TITLE
Hyperspace Fury implemented

### PR DESCRIPTION
**Pull request by MissingNoL**
## Description
Coded Hyperspace Fury by adding a BattleScript similar to Superpower's and Close Combat's and having it run from MOVE_EFFECT_FEINT.  In that, I added an "if" to check if the move is Hyperspace Fury.  Tested to properly lower Defense and break through Protect.

I think using .argument in battle_moves.h with PR #1486 could prove more effective. But I can't for the life of me understand how to get .argument and .effect to work between the PR's proposed "EFFECT_ATTACKER_DEFENSE_DOWN_HIT" and the current "EFFECT_FEINT".

## **Discord contact info**
MissingNoL#8910

### CODE REVIEW

File: src/battle_script_commands.c [hunk 1]
```diff
@@ -3148,7 +3148,12 @@ void SetMoveEffect(bool32 primary, u32 certain)
                         BattleScriptPush(gBattlescriptCurrInstr + 1);
                         gBattlescriptCurrInstr = BattleScript_MoveEffectFeint;
                     }
-                }
+				}
+				if (gCurrentMove == MOVE_HYPERSPACE_FURY)
+				{
+					BattleScriptPush(gBattlescriptCurrInstr + 1);
+					gBattlescriptCurrInstr = BattleScript_DefDown;
+				}
```
Comment (LOuroboros) [inline src/battle_script_commands.c:3156]: Spaces instead of tabs.

Also, not a fan of hardcoding the effect to Hyperspace Fury specifically like this, personally.

Comment (MissingNoL) [reply src/battle_script_commands.c:3156]: Me neither, .argument wasn't working properly though.  It would say something about there being too many characters.  Will change to spaces.

**Comment by Xhyzi** [comment]
I've noticed some a few little things that are yet to be implemented here.
- Hyperspace Fury has to be able to hit targets through substitute.
- Hyperspace Fury should only be usable by Unbound Hoopa, otherwise a "But <mon> can't use the attack!" text should be displayed.

To code the first one I'd suggest you to create a new flag called "FLAG_HIT_IN_SUBSTITUTE" or similar behind "FLAG_HIT_IN_AIR" at constants/pokemon.h. After that you only need to modify "DoesSubstituteBlockMove" function to check for that flag. (you also have to add the flag to the Hyperspace Fury movement, obv).

For the second one i'd create a "various" macro at "macro/battle_script.inc", then you can do something similar to this:

	jumpifunboundhoopa BS_ATTACKER, BattleScript_DoEffectHyperspaceFury
	printstring STRINGID_POKEMONCANNOTUSEMOVE

**Comment by LOuroboros** [comment]
> 
> 
> I've noticed a few little things that are yet to be implemented here.
> 
>     * Hyperspace Fury has to be able to hit targets through substitute.
> 
>     * Hyperspace Fury should only be usable by Unbound Hoopa, otherwise a "But  can't use the attack!" text should be displayed.
> 
> 
> To code the first one I'd suggest you to create a new flag called "FLAG_HIT_IN_SUBSTITUTE" or similar below "FLAG_HIT_IN_AIR" at constants/pokemon.h. After that you only need to modify "DoesSubstituteBlockMove" function to check for that flag. (you also have to add the flag to the Hyperspace Fury movement, obv).
> 
> For the second one i'd create a "various" macro at "macro/battle_script.inc", then you can do something similar to this:
> 
> ```
> jumpifunboundhoopa BS_ATTACKER, BattleScript_DoEffectHyperspaceFury
> printstring STRINGID_POKEMONCANNOTUSEMOVE
> ```

Personally, in regards to the BattleScript command, I believe something more general would be a better idea for futureproof's sake.

```
jumpifspecies BS_ATTACKER, SPECIES_HOOPA_UNBOUND, BattleScript_DoEffectHyperspaceFury
goto BattleScript_CannotUseMove
```

Who knows when will Game Freak decide to add more moves like this one down the road, and being prepared for that scenario sounds like a good plan.

**Comment by MissingNoL** [comment]
> > I've noticed a few little things that are yet to be implemented here.
> > ```
> > * Hyperspace Fury has to be able to hit targets through substitute.
> > 
> > * Hyperspace Fury should only be usable by Unbound Hoopa, otherwise a "But  can't use the attack!" text should be displayed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > To code the first one I'd suggest you to create a new flag called "FLAG_HIT_IN_SUBSTITUTE" or similar below "FLAG_HIT_IN_AIR" at constants/pokemon.h. After that you only need to modify "DoesSubstituteBlockMove" function to check for that flag. (you also have to add the flag to the Hyperspace Fury movement, obv).
> > For the second one i'd create a "various" macro at "macro/battle_script.inc", then you can do something similar to this:
> > ```
> > jumpifunboundhoopa BS_ATTACKER, BattleScript_DoEffectHyperspaceFury
> > printstring STRINGID_POKEMONCANNOTUSEMOVE
> > ```
> 
> Personally, in regards to the BattleScript command, I believe something more general would be a better idea for futureproof's sake.
> 
> ```
> jumpifspecies BS_ATTACKER, SPECIES_HOOPA_UNBOUND, BattleScript_DoEffectHyperspaceFury
> goto BattleScript_CannotUseMove
> ```
> 
> Who knows when will Game Freak decide to add more moves like this one down the road, and being prepared for that scenario sounds like a good plan.

I agree with more general so it can be implemented for Dark Void as well. Sorry I haven't been on this. I will also have to make a BattleScript for is the attacker is Hoopa to print a specific string (can't use in this form).

Sorry I haven't been on this! Looking for today or tomorrow to jump back on it.

**Comment by MissingNoL** [comment]
Hm when I change the species to Torchic in `data/battle_scripts_1.s` to test the different text strings, I get this build error:

`data/battle_scripts_1.o: in function `BattleScript_EffectHyperspaceFury':
(script_data+0x587): undefined reference to `SPECIES_TORCHIC'`

I don't know if this will be problematic in the future, as I added Hoopa's forms to `include/constants/battle_config.h`.  But when I add Torchic there to test the form, I also get the build errors `duplicate array index in initializer for` ___.  Not sure how I can test this otherwise, but I am very certain the code should work regardless.

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -8426,6 +8426,12 @@ static void Cmd_various(void)
             gBattlescriptCurrInstr += 7;    // exit if loop failed (failsafe)
         }
         return;
+    case VARIOUS_JUMP_IF_SPECIES:
+        if (gBattleMons[gBattlerAttacker].species == gBattlescriptCurrInstr[3])
+            gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 4);
+        else
+            gBattlescriptCurrInstr += 8;
```
Comment (LOuroboros) [inline src/battle_script_commands.c:8433]: Currently, this doesn't work. I cloned your branch and fixed it locally.
```suggestion
        if (gBattleMons[gActiveBattler].species == T1_READ_16(gBattlescriptCurrInstr + 3))
            gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 5);
        else
            gBattlescriptCurrInstr += 9;
```

**Review by LOuroboros — CHANGES_REQUESTED** [review]
